<HTML>
<HEAD>


<TITLE>Inside Java:Implementing Java Exception Handling:EarthWeb Inc.-</TITLE>






<P><CENTER>
<a href="0396-0400.html">Previous</A> | <a href="..\ewtoc.html">Table of Contents</A> | <a href="..\ch11\0409-0413.html">Next</A>
</CENTER></P>

<A NAME="PAGENUM-401"><P>Page 401</P></A>




<P>
<a href="images\ch10fg03.jpg"><img src="images\tn_ch10fg03.jpg"></a><BR>
<B>Figure 10.3</B>
<I>Output of ThrowExceptions&#151;Screen 1.</I>
</P>



<P>
<a href="images\ch10fg04.jpg"><img src="images\tn_ch10fg04.jpg"></a><BR>
<b>Figure 10.4</b>
<I>Output of ThrowExceptions&#151;Screen 2.
</I>

<P>Listing 10.1
</P>
<PRE>
Line==================Java Source=================
001 /**
002  * ThrowExceptions: Demonstrates the use of
003  * exceptions:
004  *
005  * Author: Karanjit S. G. S. Siyan
006  *
007  */
</PRE>
<PRE>
								<I>continues</I>
</PRE>

<A NAME="PAGENUM-402"><P>Page 402</P></A>




<P>Listing 10.1, Continued
</P>
<PRE>
008
009 // IntException class
010 class IntException extends Exception
011 {
012
013     // Constructor with one arg
014     IntException(String msg)
015     {
016         super(msg);
017
018     }  // end IntException
019
020     // Constructor with no args
021     // Default constructor
022     IntException()
023     {
024         super();
025
026     }  // end IntException
027
028 }  // end IntException
029
030 // A subclass of the IntException class
031 class IntSubException extends IntException
032 {
033
034     protected String excepMsg;
035
036     // Constructor with one arg
037     IntSubException(String msg)
038     {
039         super(msg);
040         excepMsg = msg;
041
042     }  // end IntSubException
043
044     // Constructor with no args
045     // Default constructor
</PRE>


<A NAME="PAGENUM-403"><P>Page 403</P></A>


<PRE>
046     IntSubException()
047     {
048         super();
049         excepMsg = &quot;thrown without message&quot;;
050
051     }  // end IntSubException
052
053     // Overriding toString() in Throwable
054     public String toString()
055     {
056         return &quot;{IntSubException: &quot;+excepMsg+&quot;}&quot;;
057
058     }  // end toString
059
060 }  // end IntSubException
061
062 // Another exception class
063 class IntLargeException extends Exception
064 {
065
066     // Constructor with one arg
067     IntLargeException(String msg)
068     {
069         super(msg);
070
071     }  // end IntLargeException
072
073     // Constructor with no args
074     // Default constructor
075     IntLargeException()
076     {
077         super();
078
079     }  // end IntLargeException
080
081 }  // end IntLargeException
082
083
</PRE>
<PRE>
								<I>continues</I>
</PRE>

<A NAME="PAGENUM-404"><P>Page 404</P></A>




<P>Listing 10.1, Continued
</P>
<PRE>
084 public class ThrowExceptions
085 {
086
087     public static void main(String[] args)
088     {
089         int     iarg;           // Integer arg
090
091         // Parse the passed parameter
092         try
093         {
094             iarg = Integer.parseInt(args[0]);
095         }
096         catch (ArrayIndexOutOfBoundsException ae)
097         {
098             // Arg not supplied
099             System.out.println(&quot;Must specify at least one argument&quot;);
100             System.out.println(&quot;Usage:\n\tjava ThrowExceptions intval\n&quot;);
101             return;
102
103         }
104         catch (NumberFormatException ne)
105         {
106             // Expecting an integer
107             System.out.println(&quot;Must specify an int value&quot;);
108             System.out.println(&quot;Usage:\n\tjava ThrowExceptions intval\n&quot;);
109             return;
110
111         }  // end try
112
113         // Print the square root.
114         takeSqrt(iarg);
115
116     }  // end main
117
118     // Prints square root if number is
119     // not odd and a perfect square.
120     public static void takeSqrt(int ival)
121     {
</PRE>

<A NAME="PAGENUM-405"><P>Page 405</P></A>



<PRE>
122
123         // First check the integer value
124         try
125         {
126             chkIntVal(ival);
127         }
128         catch (IntException ie)
129         {
130             if (ie instanceof IntSubException)
131             {
132                 System.out.println(&quot;IntSubException thrown&quot;);
133             }
134             else
135             {
136                 System.out.println(&quot;IntException thrown&quot;);
137
138             }  // end if
139
140             System.out.println(&quot;\ttoString()=&quot;+ie.toString());
141             System.out.println(&quot;\tprintln()=&quot;+ie);
142             ie.printStackTrace();
143             return;
144
145         }
146         catch  (IntLargeException ie)
147         {
148             System.out.println(&quot;IntException thrown&quot;);
149             System.out.println(&quot;\ttoString()=&quot;+ie.toString());
150             System.out.println(&quot;\tprintln()=&quot;+ie);
151             ie.printStackTrace();
152             return;
153
154         }  // end try
155
156         // Prints the square root.
157         System.out.println(&quot;Square root = &quot;+
158             (int)Math.sqrt((double)ival));
159
</PRE>
<PRE>
								<I>continues</I>
</PRE>

<A NAME="PAGENUM-406"><P>Page 406</P></A>




<P>Listing 10.1, Continued
</P>
<PRE>
160     }  // end takeSqrt
161
162     // Checks the integer value.
163     private static void chkIntVal(int ival)
164         throws IntException, IntLargeException,
165                IntSubException
166     {
167
168         int[] perfectSqs = {0, 4, 16, 36,  64, 100,
169                             144, 196, 256, 324, 400,
170                             484, 576, 676, 784, 900};
171
172         if (ival &lt; 0)
173         {
174             throw new IntException(&quot;Negative value for sqrt!&quot;);
175
176         }
177         else if (ival &gt; 1000)
178         {
179             throw new IntLargeException(&quot;Number greater than 1000&quot;);
180
181         }  // end if
182
183
184         if (ival  == 1)
185         {
186             throw new IntSubException(&quot;Odd number&quot;);
187
188         }  // end if
189
190
191         for (int x = 0; x &lt; perfectSqs.length; x++)
192         {
193             if (perfectSqs[x] == ival)
194             {
195                 return;
196
197             } // end if
</PRE>


<A NAME="PAGENUM-407"><P>Page 407</P></A>


<PRE>
198
199         } // end for
200
201         throw new IntSubException(ival+&quot; not a perfect square&quot;);
202
203     }  // end chkIntVal
204
205 }  // end ThrowExceptions
</PRE>

<H3><A NAME="ch10_ 10">
Summary
</A></H3>

<P>Java provides a robust method for handling error conditions by providing
language constructs for handling errors as exceptions. You can define your own exceptions
by extending the Throwable or Exception class already provided as part of the
java.lang package.
</P>

<P>In this chapter, you learned how exceptions are defined and how they are thrown.
You learned how to use the catch statement to catch exceptions and the difference
between checked and unchecked exceptions. Most exceptions that you define in your
program will be checked exceptions. Unchecked exceptions are usually runtime exceptions
or severe errors and can occur anywhere within a program.
</P>

<A NAME="PAGENUM-408"><P>Page 408</P></A>



<P><CENTER>
<a href="0396-0400.html">Previous</A> | <a href="..\ewtoc.html">Table of Contents</A> | <a href="..\ch11\0409-0413.html">Next</A>
</CENTER></P>







</BODY>
</HTML>

